LiskHQ/lisk-mobile
# Update React Navigation

# What was the bug or feature?
**Important: Rebased #902 - Please review and merge #902 before merging this PR.**

Described in #897.

### How did I fix it?
- Upgraded React Navigation to v2.18.
- Replaced `navigate` with `push` where needed to comply with [new navigation functionality](https://reactnavigation.org/blog/2018/05/07/react-navigation-2.0.html#navigateroutename-in-stacknavigator-is-less-pushy).
- Rebased #902 to make use of newer version of Gradle.
- Installed [React Navigation Tabs](https://github.com/react-navigation/tabs) to replace deprecated `TabBarComponent`.
- Installed [react-native-gesture-handler](https://github.com/kmagiera/react-native-gesture-handler) and [react-native-reanimated](https://github.com/kmagiera/react-native-reanimated) as part of React Native Tabs installation.
- Installed peer dependencies [@react-navigation/core](https://github.com/react-navigation/core) and [@react-navitaion/native](https://github.com/react-navigation/native).
- Refactored `navigationOptions`.
- Moved `Send` and `Home` to their own stack navigator due to new [shallow navigation](https://reactnavigation.org/blog/2018/05/07/react-navigation-2.0.html#shallow-navigation-options).
- Removed `tabBar` on Android when sticky button is on keyboard to avoid buggy behaviour of keyboard - behaves same as iOS now.

TODOS: 
- [x] Fix back button on `Send`.
- [x] Fix header animation on `Home`.

## Type of change
- Enhancement (a non-breaking change which adds functionality)

### How to test it?
- Make sure that navigation works on all screens and that no functionality was broken.
- Test that it is possible to push commits to remote. Sometimes some of the jest tests have failed for no reason and `node_modules` need to be removed in order to push. Seems to be caused by the changes on #902.
- Test the `Send` flow on iPhone X or superior. The sticky button has sometimes a slightly inconsistent height and behaves differently depending on which button you navigate from - please alternate between sticky button and any other navigation button on the screen when testing. Inputting something on the keyboard sometimes slightly changes the height too and most fixes affect the layout either on other devices or on other instances of the button. I recommend creating a separate issue to fix this. Consider replacing KeyboardAccessoryView with React Native's [InputAccessoryView](https://facebook.github.io/react-native/docs/0.59/inputaccessoryview).
- Test fails because of `rn-nodeify` not found (?) - could be related to gradle.

# Checklist:
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- I have added tests that prove my fix is effective or that my feature works
- New and existing unit tests pass locally with my changes
